raiden_contracts/data/source/raiden/TokenNetwork.sol

```diff
@@ -479,7 +479,77 @@ contract TokenNetwork is Utils {
         assert(participant_state.nonce == 0);
         assert(partner_state.nonce == 0);
+    }
+
+    struct CoopData {
+        address participant;
+        uint256 total_withdraw;
+        uint256 expiration_block;
+        bytes participant_signature;
+        bytes partner_signature;
+    }
+
+    function cooperativeSettle(
+        uint256 channel_identifier,
+        CoopData memory data1,
+        CoopData memory data2
+    )
+        external
+        isOpen(channel_identifier)
+    {
+        uint256 total_deposit;
+        bytes32 pair_hash;
+
+        // Validate that authenticated partners and the channel identifier match
+        require(channel_identifier == getChannelIdentifier(data1.participant, data2.participant));
+        Channel storage channel = channels[channel_identifier];
+
+        Participant storage participant1_state = channel.participants[data1.participant];
+        Participant storage participant2_state = channel.participants[data2.participant];
+        total_deposit = participant1_state.deposit + participant2_state.deposit;
+
+        // The sum of the provided balances must be equal to the total
+        // available deposit
+        require((data1.total_withdraw + data2.total_withdraw) == total_deposit);
+        // Overflow check for the balances addition from the above check.
+        // This overflow should never happen if the token.transfer function is implemented
+        // correctly. We do not control the token implementation, therefore we add this
+        // check for safety.
+        require(data1.total_withdraw <= data1.total_withdraw + data2.total_withdraw);
+
+        if (data1.total_withdraw > 0) {
+            this.setTotalWithdraw(
+                channel_identifier,
+                data1.participant,
+                data1.total_withdraw,
+                data1.expiration_block,
+                data1.participant_signature,
+                data1.partner_signature
+            );
+        }
+        if (data2.total_withdraw > 0) {
+            this.setTotalWithdraw(
+                channel_identifier,
+                data2.participant,
+                data2.total_withdraw,
+                data2.expiration_block,
+                data2.participant_signature,
+                data2.partner_signature
+            );
+        }
+
+        // Remove channel data from storage
+        delete channel.participants[data1.participant];
+        delete channel.participants[data2.participant];
+        delete channels[channel_identifier];
+
+        // Remove the pair's channel counter
+        pair_hash = getParticipantsHash(data1.participant, data2.participant);
+        delete participants_hash_to_channel_identifier[pair_hash];
 
+        // Emit channel lifecycle events
+        emit ChannelClosed(channel_identifier, data1.participant, 0, 0);
+        emit ChannelSettled(channel_identifier, data1.total_withdraw, 0, data2.total_withdraw, 0);
```

Comment:
Should we reuse the code from https://github.com/raiden-network/raiden-contracts/blob/15037b89cf322d1a1d309d097e06b7ed72c9e0b8/raiden_contracts/data/source/raiden/TokenNetwork.sol#L782-L809 here?

This question also shows that we are missing `storeUnlockData`. Is this a bug, or is there an explanation why we don't need it? We should add an explanation (if it is not needed) or a test to unlock after settlement (if it is needed).

Reply:
I factored out the mentioned lines and added some comments about why this doesn't allow remaining locks in the docstring as well as the code.